internal/command/show.go

```diff
@@ -24,243 +23,228 @@ type ShowCommand struct {
+			return plan, stateFile, config, schemas, diags
 		}
-		c.Ui.Output(string(jsonState))
-	} else {
-		if stateFile == nil {
-			c.Ui.Output("No state.")
-			return 0
+		tfCtx, ctxDiags := terraform.NewContext(opts)
+		diags = diags.Append(ctxDiags)
+		if ctxDiags.HasErrors() {
+			return plan, stateFile, config, schemas, diags
+		}
+		var schemaDiags tfdiags.Diagnostics
+		schemas, schemaDiags = tfCtx.Schemas(config, stateFile.State)
+		diags = diags.Append(schemaDiags)
+		if schemaDiags.HasErrors() {
+			return plan, stateFile, config, schemas, diags
 		}
-		c.Ui.Output(format.State(&format.StateOpts{
-			State:   stateFile.State,
-			Color:   c.Colorize(),
-			Schemas: schemas,
-		}))
 	}
 
-	return 0
+	return plan, stateFile, config, schemas, diags
 }
+func (c *ShowCommand) showFromLatestStateSnapshot() (*statefile.File, tfdiags.Diagnostics) {
+	var diags tfdiags.Diagnostics
 
-func (c *ShowCommand) Help() string {
-	helpText := `
-Usage: terraform [global options] show [options] [path]
-
-  Reads and outputs a Terraform state or plan file in a human-readable
-  form. If no path is specified, the current state will be shown.
+	// Load the backend
+	b, backendDiags := c.Backend(nil)
+	diags = diags.Append(backendDiags)
+	if backendDiags.HasErrors() {
+		return nil, diags
+	}
+	c.ignoreRemoteVersionConflict(b)
 
-Options:
+	// Load the workspace
+	workspace, err := c.Workspace()
+	if err != nil {
+		diags = diags.Append(fmt.Errorf("error selecting workspace: %s", err))
+		return nil, diags
+	}
 
-  -no-color           If specified, output won't contain any color.
-  -json               If specified, output the Terraform plan or state in
-                      a machine-readable form.
+	// Get the latest state snapshot from the backend for the current workspace
+	stateFile, stateErr := getStateFromBackend(b, workspace)
+	if stateErr != nil {
+		diags = diags.Append(stateErr.Error())
+		return nil, diags
+	}
 
-`
-	return strings.TrimSpace(helpText)
+	return stateFile, diags
 }
 
-func (c *ShowCommand) Synopsis() string {
-	return "Show the current state or a saved plan"
+func (c *ShowCommand) showFromPath(path string) (*plans.Plan, *statefile.File, *configs.Config, tfdiags.Diagnostics) {
+	var diags tfdiags.Diagnostics
+	var planErr, stateErr error
+	var plan *plans.Plan
+	var stateFile *statefile.File
+	var config *configs.Config
+
+	// Try to get the plan file and associated data from
+	// the path argument. If that fails, try to get the
+	// statefile from the path argument.
+	plan, stateFile, config, planErr = getPlanFromPath(path)
+	if planErr != nil {
+		stateFile, stateErr = getStateFromPath(path)
+		if stateErr != nil {
+			diags = diags.Append(fmt.Errorf(
+				"Terraform couldn't read the given file as a state or plan file.\n"+
+					"The errors while attempting to read the file as each format are\n"+
+					"shown below.\n\n"+
+					"State read error: %s\n\nPlan read error: %s",
+				stateErr,
+				planErr))
```

Comment:
This err-as-diagnostic looks really awkward in the error output:

<img width="793" alt="image" src="https://user-images.githubusercontent.com/68917/149362981-fc82b168-2c71-4b30-a439-5cd4aa75a596.png">

We could use a sourceless diagnostic here instead to improve the formatting a little. For example:

<img width="724" alt="image" src="https://user-images.githubusercontent.com/68917/149363558-79057557-6204-4a90-b424-654002636fc3.png">

```go
			diags = diags.Append(tfdiags.Sourceless(
				tfdiags.Error,
				"Failed to read the given file as a state or plan file",
				fmt.Sprintf(
					"State read error: %s\n\nPlan read error: %s",
					stateErr,
					planErr,
				),
			))
```

Reply:
done!